### PR TITLE
v4 api enhancements

### PR DESCRIPTION
**Pull request by gnodet**
- Fix DefaultArtifactManager to allow using both Artifact and Dependency
- Add a TODO for an EventSpy related api
- Add simple string/model conversion on the XmlFactory
- Decouple services creation

Following this checklist to help us incorporate your
contribution quickly and easily:

 - [ ] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/MNG) filed
       for the change (usually before you start working on it).  Trivial changes like typos do not
       require a JIRA issue. Your pull request should address just this issue, without
       pulling in other changes.
 - [ ] Each commit in the pull request should have a meaningful subject line and body.
 - [ ] Format the pull request title like `[MNG-XXX] SUMMARY`, where you replace `MNG-XXX`
       and `SUMMARY` with the appropriate JIRA issue. Best practice is to use the JIRA issue
       title in the pull request title and in the first line of the commit message.
 - [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [ ] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will
       be performed on your pull request automatically.
 - [ ] You have run the [Core IT][core-its] successfully.

If your pull request is about ~20 lines of code you don't need to sign an
[Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf) if you are unsure
please ask on the developers list.

To make clear that you license your contribution under
the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
you have to acknowledge this by using the following check-box.

 - [ ] I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)

 - [ ] In any other case, please file an [Apache Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

[core-its]: https://maven.apache.org/core-its/core-it-suite/

### CODE REVIEW

File: maven-core/src/main/java/org/apache/maven/internal/impl/DefaultArtifactInstaller.java [hunk 1]
```diff
@@ -38,11 +41,13 @@
 import static org.apache.maven.internal.impl.Utils.cast;
 import static org.apache.maven.internal.impl.Utils.nonNull;
 
+@Named
```
Comment (cstamas) [inline maven-core/src/main/java/org/apache/maven/internal/impl/DefaultArtifactInstaller.java:44]: and not singleton?

File: maven-core/src/main/java/org/apache/maven/internal/impl/DefaultArtifactResolver.java [hunk 2]
```diff
@@ -43,6 +44,7 @@
 import static org.apache.maven.internal.impl.Utils.cast;
 import static org.apache.maven.internal.impl.Utils.nonNull;
 
+@Named
```
Comment (cstamas) [inline maven-core/src/main/java/org/apache/maven/internal/impl/DefaultArtifactResolver.java:47]: singleton?

**Comment by cstamas** [comment]
general remark: all those "named" classes are not singletons? is this on purpose?

**Comment by gnodet** [comment]
> general remark: all those "named" classes are not singletons? is this on purpose?

They are stateless beans, so they could be moved to the `@Singleton` scope.  The question is: should they ?

**Comment by cstamas** [comment]
> They are stateless beans, so they could be moved to the `@Singleton` scope. The question is: should they ?

In general, components should be stateless and immutable. By having them in singleton scope, in some circumstances you can preserve some memory and/or GC pressure, especially if instances are used across multiple components (is enlisted several times in object graph) or, (this is more like webapp scenario) are injected per request or so. 

Maven codebase in general applies this clearly and consistently, as IMHO the intent is clear when you see singleton annotation on a component... So my preference is yes, if a component is meant to be singleton, then they should be made singletons (as later on, as codebase evolve, this can lead to issues, IMHO be explicit and follow Maven "style" re components)

**Review by cstamas — APPROVED** [review]

**Pull request merged**